src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+    let fatalError = false;
+
+    if (!addresses) {
+      await req.scope.cradle.api.getCourtAddresses(slug)
+        .then((addressList: CourtAddress[]) => addresses = this.convertToDisplayAddresses(addressList))
+        .catch(() => {
+          errorMsgs.push(this.getAddressesError);
+          fatalError = true;
+        });
+    }
+
+    let addressTypes: AddressType[] = [];
+    await req.scope.cradle.api.getAddressTypes()
+      .then((types: AddressType[]) => addressTypes = types)
+      .catch(() => {
+        errorMsgs.push(this.getAddressTypesError);
+        fatalError = true;
+      });
+    const writeToUsTypes = addressTypes.filter(at => at.name.toLowerCase() === this.writeToUsAddressType.toLowerCase()).map(at => at.id);
+
+    const pageData: CourtAddressPageData = {
+      addressTypesPrimary: this.getAddressTypesForSelect(addressTypes, true),
+      addressTypesSecondary: this.getAddressTypesForSelect(addressTypes, false),
+      addresses: addresses,
+      writeToUsTypeId: writeToUsTypes.length === 1 ? writeToUsTypes[0] : null,
+      errors: errorMsgs.map(errorMsg => { return { text: errorMsg }; }),
+      fatalError: fatalError,
+      primaryPostcodeInvalid: primaryPostcodeInvalid,
+      secondaryPostcodeInvalid: secondaryPostcodeInvalid,
+      updated: updated
+    };
+
+    res.render('courts/tabs/addressesContent', pageData);
+  }
+
+  private validateAddress(address: DisplayAddress, isPrimaryAddress: boolean): AddressValidationResult {
+    const typeErrors = this.validateType(address, isPrimaryAddress);
+    const addressErrors = this.validateAddressLines(address, isPrimaryAddress);
+    const postcodeErrors = this.validatePostcode(address, isPrimaryAddress);
+    const errorPrefix = isPrimaryAddress ? this.primaryAddressPrefix : this.secondaryAddressPrefix;
+
+    return {
+      postcodeValid: postcodeErrors.length === 0,
+      addressValid: addressErrors.length === 0,
+      errors: typeErrors.concat(addressErrors).concat(postcodeErrors).map(error => errorPrefix + error)
+    };
+  }
+
+  private validateType(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    if (isPrimaryAddress || (!!address.postcode || this.addressFieldsNotEmpty(address))) {
+      return !address.type_id ? [this.typeRequiredError] : [];
+    }
+    return [];
+  }
+
+  private validateAddressLines(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const validateSecondaryAddress = (!isPrimaryAddress && (this.addressFieldsNotEmpty(address) || address.postcode?.trim()));
+    const errors: string[] = [];
+
+    if (isPrimaryAddress || validateSecondaryAddress) {
+      if (!address.address_lines?.trim()) {
+        errors.push(this.addressRequiredError);
+      }
+      if(!address.town?.trim()) {
+        errors.push(this.townRequiredError);
+      }
+    }
+    return errors;
+  }
+
+  private validatePostcode(courtAddress: DisplayAddress, isPrimaryAddress: boolean): string[] {
```

Comment:
Only double checking on this one, so feel free to resolve if its true. The assumption is that the regex check for a postcode on an address will happen on the API from memory, and we are not doing the same on the frontend? 

Benefit I see to doing so is saving a network call, but I understand or I think that we don't do that for the others